### PR TITLE
Added getMasternodeBlocks

### PR DESCRIPTION
**Pull request by siradji**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:
Added getMasternodeBlocks rpc
#### Which issue(s) does this PR fixes?:

Fixes part of #48

### CODE REVIEW

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+    const blocksResult = await client.masternode.getMasternodeBlocks({ ownerAddress: address })
+
+    for (const value of Object.values(blocksResult)) {
+      expect(typeof value).toStrictEqual('string')
+      expect(value.length).toStrictEqual(64)
+    }
+  })
+
+  it('should getMasternodeBlocks with depth', async () => {
```
Comment (canonbrother) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:46]: what's the diff of getting master node with or without depth??

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    const blocksResult = await client.masternode.getMasternodeBlocks({ id: masternodeID })
+
+    for (const value of Object.values(blocksResult)) {
+      expect(typeof value).toStrictEqual('string')
+      expect(value.length).toStrictEqual(64)
+    }
+  })
+
+  it('should getMasternodeBlocks with owner address', async () => {
```
Comment (canonbrother) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:32]: missing operator address??

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  it('should getMasternodeBlocks', async () => {
+    const masternodeID = await client.masternode.createMasternode(await client.wallet.getNewAddress())
+
+    await container.generate(1)
+
+    const blocksResult = await client.masternode.getMasternodeBlocks({ id: masternodeID })
+
+    for (const value of Object.values(blocksResult)) {
+      expect(typeof value).toStrictEqual('string')
+      expect(value.length).toStrictEqual(64)
```
Comment (canonbrother) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:28]: hmm.. how to get use of the string??

**Comment by canonbrother** [comment]
> #### What kind of PR is this?:
> 
> /kind feature
> #### What this PR does / why we need it:
> 
> rpc_masternodes.cpp implementation #48
> #### Which issue(s) does this PR fixes?:
> 
> Fixes part of #48
> #### Additional comments?:

can you follow the rpc pull request template format  as usual??

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts (hunk 3 above)
Comment (siradji) [reply packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:28]: The string is the block hash returned

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts (hunk 1 above)
Comment (siradji) [reply packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:46]: No difference. I checked

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts (hunk 2 above)
Comment (siradji) [reply packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:32]: ownerAddress and operatorAddress are interchangeable that is why i felt it is redundant to add a separate test

**Comment by jingyi2811** [comment]
You do not include operatorAddress in your tests
See https://github.com/DeFiCh/jellyfish/pull/418/files#diff-4d7e75f85e53bcd52f372263dedc6e010c564ed07bf046faf3f5024704a08ac3R161

I would recommend you redesign your tests like this

Pass in Id only, what happened?
Pass in ownerAddress only, what happened?
Pass in operatorAddress only, what happened?

Pass in invalid Id only, what happened?
Pass in invalid ownerAddress only, what happened?
Pass in invalid operatorAddress only, what happened?

Pass in Id, ownerAddress, operatorAddress, what happened?
Pass in id and depth, what happened?

**Comment by siradji** [comment]
> You do not include operatorAddress in your tests
> See https://github.com/DeFiCh/jellyfish/pull/418/files#diff-4d7e75f85e53bcd52f372263dedc6e010c564ed07bf046faf3f5024704a08ac3R161
> 
> I would recommend you redesign your tests like this
> 
> Pass in Id only, what happened?
> Pass in ownerAddress only, what happened?
> Pass in operatorAddress only, what happened?
> 
> Pass in invalid Id only, what happened?
> Pass in invalid ownerAddress only, what happened?
> Pass in invalid operatorAddress only, what happened?
> 
> Pass in Id, ownerAddress, operatorAddress, what happened?
> Pass in id and depth, what happened?

@jingyi2811  added additional test to cover what you outlined. Thanks!

**Review by canonbrother — APPROVED** [review]

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    await container.waitForReady()
+    await container.waitForWalletCoinbaseMaturity()
+  })
+
+  afterAll(async () => {
+    await container.stop()
+  })
+
+  it('should getMasternodeBlocks', async () => {
+    const masternodeID = await client.masternode.createMasternode(await client.wallet.getNewAddress())
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:20]: Use `container.call` like what Webster mentioned in your another PR

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+    const blocksResult = await client.masternode.getMasternodeBlocks({ operatorAddress: address })
+
+    for (const value of Object.values(blocksResult)) {
+      expect(typeof value).toStrictEqual('string')
+      expect(value.length).toStrictEqual(64)
+    }
+  })
+
+  it('should getMasternodeBlocks with depth and owner address', async () => {
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:60]: Change the text to `should getMasternodeBlocks with owner address and depth`

Owner address = 1st param
Depth = 2nd param

Apply to other text too

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    await container.start()
+    await container.waitForReady()
+    await container.waitForWalletCoinbaseMaturity()
+  })
+
+  afterAll(async () => {
+    await container.stop()
+  })
+
+  it('should getMasternodeBlocks', async () => {
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:19]: `should getMasternodeBlocks with id`

**Comment by jingyi2811** [comment]
Also, rearrange your test in this order => id, owner address, operator address

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts (hunk 4 above)
Comment (siradji) [reply packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:20]: Remember `container.call` fails on masternode rpcs? The only option here is to use `container.getnewaddress`

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    await container.waitForReady()
+    await container.waitForWalletCoinbaseMaturity()
+  })
+
+  afterAll(async () => {
+    await container.stop()
+  })
+
+  it('should getMasternodeBlocks with ID', async () => {
+    const masternodeID = await client.masternode.createMasternode(await container.getNewAddress())
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:20]: Let's make this easier to be read

`const id`

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+    await container.generate(1)
+
+    const blocksResult = await client.masternode.getMasternodeBlocks({ id: masternodeID })
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:24]: const blocksResult = await client.masternode.getMasternodeBlocks({ id })

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    await container.start()
+    await container.waitForReady()
+    await container.waitForWalletCoinbaseMaturity()
+  })
+
+  afterAll(async () => {
+    await container.stop()
+  })
+
+  it('should getMasternodeBlocks with ID', async () => {
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:19]: `should getMasternodeBlocks with id`. ID should be lowercase like how you declare in the rest of the tests

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    const blocksResult = await client.masternode.getMasternodeBlocks({ id: masternodeID })
+
+    for (const value of Object.values(blocksResult)) {
+      expect(typeof value).toStrictEqual('string')
+      expect(value.length).toStrictEqual(64)
+    }
+  })
+
+  it('should getMasternodeBlocks with owner address', async () => {
+    const address = await container.getNewAddress()
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:33]: const ownerAddress

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+    await client.masternode.createMasternode(address)
+
+    await container.generate(1)
+
+    const blocksResult = await client.masternode.getMasternodeBlocks({ ownerAddress: address })
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:38]: `const blocksResult = await client.masternode.getMasternodeBlocks({ ownerAddress })`

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+    }
+  })
+
+  it('should getMasternodeBlocks with operator address', async () => {
+    const address = await container.getNewAddress()
+    await client.masternode.createMasternode(address)
+
+    await container.generate(1)
+
+    const blocksResult = await client.masternode.getMasternodeBlocks({ operatorAddress: address })
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:52]: Same

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+    const blocksResult = await client.masternode.getMasternodeBlocks({ ownerAddress: address })
+
+    for (const value of Object.values(blocksResult)) {
+      expect(typeof value).toStrictEqual('string')
+      expect(value.length).toStrictEqual(64)
+    }
+  })
+
+  it('should getMasternodeBlocks with operator address', async () => {
+    const address = await container.getNewAddress()
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:47]: `operatorAddress`

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+    const blocksResult = await client.masternode.getMasternodeBlocks({ operatorAddress: address })
+
+    for (const value of Object.values(blocksResult)) {
+      expect(typeof value).toStrictEqual('string')
+      expect(value.length).toStrictEqual(64)
+    }
+  })
+
+  it('should getMasternodeBlocks with id and depth', async () => {
+    const masternodeId = await client.masternode.createMasternode(await container.getNewAddress())
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:61]: `id`

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+
+    await container.generate(1)
+
+    const blocksResult = await client.masternode.getMasternodeBlocks({ id: masternodeId }, 10)
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:65]: Same

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+
+    for (const value of Object.values(blocksResult)) {
+      expect(typeof value).toStrictEqual('string')
+      expect(value.length).toStrictEqual(64)
+    }
+  })
+
+  it('should getMasternodeBlocks with owner address and depth', async () => {
+    const address = await container.getNewAddress()
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:74]: Same

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+    await client.masternode.createMasternode(address)
+
+    await container.generate(1)
+
+    const blocksResult = await client.masternode.getMasternodeBlocks({ ownerAddress: address }, 10)
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:79]: Same

Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:79]: And apply to all too

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+    const blocksResult = await client.masternode.getMasternodeBlocks({ operatorAddress }, 10)
+
+    for (const value of Object.values(blocksResult)) {
+      expect(typeof value).toStrictEqual('string')
+      expect(value.length).toStrictEqual(64)
+    }
+  })
+
+  it('should throw and error with invalid id', async () => {
+    const invalidId = 'c3285ad8ab28886beeeeb92562f40168ba8d877cff6f2f02301c99053ed33349'
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:102]: id

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+    const promise = client.masternode.getMasternodeBlocks({ id: invalidId })
+
+    await expect(promise).rejects.toThrow(RpcApiError)
+    await expect(promise).rejects.toThrow('Masternode not found')
+  })
+
+  it('should throw an error with invalid owner address ', async () => {
+    const invalidOwnerAddress = 'INVALIDADDRESS'
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:110]: ownerAddress

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+    const promise = client.masternode.getMasternodeBlocks({ ownerAddress: invalidOwnerAddress })
+
+    await expect(promise).rejects.toThrow(RpcApiError)
+    await expect(promise).rejects.toThrow('Invalid P2PKH address')
+  })
+
+  it('should throw an error with invalid operator address ', async () => {
+    const invalidOperatorAddress = 'INVALIDADDRESS'
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:118]: operatorAddress

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+
+  it('should throw an error when identifier is not passed', async () => {
+    const promise = client.masternode.getMasternodeBlocks({})
+
+    await expect(promise).rejects.toThrow(RpcApiError)
+    await expect(promise).rejects.toThrow('Provide masternode identifier information')
+  })
+
+  it('should throw an error when multiple identifiers are provided', async () => {
+    const ownerAddress = await container.getNewAddress()
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:133]: operatorAddress

Comment (siradji) [reply packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:133]: We have to use the ownerAddress there as operator address.

**Review by jingyi2811 — DISMISSED** [review]

**Comment by fuxingloh** [comment]
@siradji conflicts

**Comment by siradji** [comment]
> @siradji conflicts

@fuxingloh, all conflicts are resolved. Thanks!

**Review by jingyi2811 — APPROVED** [review]

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+
+  it('should getMasternodeBlocks with operator address and depth', async () => {
+    const operatorAddress = await container.getNewAddress()
+    await client.masternode.createMasternode(operatorAddress)
+
+    await container.generate(1)
+
+    const blocksResult = await client.masternode.getMasternodeBlocks({ operatorAddress }, 10)
+
+    for (const value of Object.values(blocksResult)) {
```
Comment (fuxingloh) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:95]: assert result size

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+
+  it('should getMasternodeBlocks with owner address and depth', async () => {
+    const ownerAddress = await container.getNewAddress()
+    await client.masternode.createMasternode(ownerAddress)
+
+    await container.generate(1)
+
+    const blocksResult = await client.masternode.getMasternodeBlocks({ ownerAddress }, 10)
+
+    for (const value of Object.values(blocksResult)) {
```
Comment (fuxingloh) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:81]: assert result size

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+  })
+
+  it('should getMasternodeBlocks with id and depth', async () => {
+    const id = await client.masternode.createMasternode(await container.getNewAddress())
+
+    await container.generate(1)
+
+    const blocksResult = await client.masternode.getMasternodeBlocks({ id }, 10)
+
+    for (const value of Object.values(blocksResult)) {
```
Comment (fuxingloh) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:67]: assert result size

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 25]
```diff
@@ -0,0 +1,143 @@
...
+
+  it('should getMasternodeBlocks with operator address', async () => {
+    const operatorAddress = await container.getNewAddress()
+    await client.masternode.createMasternode(operatorAddress)
+
+    await container.generate(1)
+
+    const blocksResult = await client.masternode.getMasternodeBlocks({ operatorAddress })
+
+    for (const value of Object.values(blocksResult)) {
```
Comment (fuxingloh) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:54]: assert result size

File: packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts [hunk 26]
```diff
@@ -0,0 +1,143 @@
...
+
+  it('should getMasternodeBlocks with owner address', async () => {
+    const ownerAddress = await container.getNewAddress()
+    await client.masternode.createMasternode(ownerAddress)
+
+    await container.generate(1)
+
+    const blocksResult = await client.masternode.getMasternodeBlocks({ ownerAddress })
+
+    for (const value of Object.values(blocksResult)) {
```
Comment (fuxingloh) [inline packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlock.test.ts:40]: assert result size

**Review by fuxingloh — CHANGES_REQUESTED** [review]
missing size assertions when you query

I believe you can:
`container.generate()` to another master node automatically

via the use of this method: 
https://github.com/DeFiCh/jellyfish/blob/490f21af970a518097f77c2fcd473c66c3f14fd5/packages/testcontainers/src/chains/reg_test_container/masternode.ts#L37

**Comment by fuxingloh** [comment]
> missing size assertions when you query
> 
> I believe you can:
> `container.generate()` to another master node automatically
> 
> via the use of this method:
> https://github.com/DeFiCh/jellyfish/blob/490f21af970a518097f77c2fcd473c66c3f14fd5/packages/testcontainers/src/chains/reg_test_container/masternode.ts#L37

Impossible to test without https://github.com/DeFiCh/jellyfish/issues/40 as you need to generate a block to another operator.

**Comment by fuxingloh** [comment]
> > missing size assertions when you query
> > I believe you can:
> > `container.generate()` to another master node automatically
> > via the use of this method:
> > https://github.com/DeFiCh/jellyfish/blob/490f21af970a518097f77c2fcd473c66c3f14fd5/packages/testcontainers/src/chains/reg_test_container/masternode.ts#L37
> 
> Impossible to test without #40 as you need to generate a block to another operator.

https://github.com/DeFiCh/jellyfish/pull/450/files#diff-4f8572804c2157e6bd50186d19c4ac65ece099e1cb04725afc35e0d16e3e6ffeR42

**Pull request closed without merging**